src/main/rust/src/fs_monitor.rs

```diff
@@ -0,0 +1,153 @@
+}
+
+impl NativeEventStream {
+    pub fn new(path: String, handler: impl Fn(Kind, &String) + 'static) -> Self {
+        Self {
+            since_when: unsafe { fse::FSEventsGetCurrentEventId() },
+            closed: AtomicBool::new(false),
+            path: CFArray::from_CFTypes(&[CFString::new(&path)]),
+            queue: unsafe { dispatch_queue_create(ptr::null(), DISPATCH_QUEUE_SERIAL) },
+            stream: None,
+            info: Box::into_raw(Box::new(ContextInfo{ handler: Box::new(handler) })),
+        }
+    }
+
+    pub fn start(&mut self) {
+        unsafe {
+            let stream = fse::FSEventStreamCreate(
+                kCFAllocatorDefault,
+                callback,
+                &fse::FSEventStreamContext {
+                    version: 0,
+                    info: self.info as *mut _,
+                    retain: None,
+                    release: Some(release_context),
+                    copy_description: None
+                },
+                self.path.as_concrete_TypeRef(),
+                self.since_when,
+                0.15,
+                FLAGS);
+
+            self.stream = Some(stream);
+
+            fse::FSEventStreamSetDispatchQueue(stream, self.queue);
+            fse::FSEventStreamStart(stream);
+        };
+    }
+
+    pub fn stop(&self) {
+        if self.closed.compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst).is_err() {
+            return; // The stream has already been closed
+        }
+        match self.stream {
+            Some(stream) => unsafe{
+                fse::FSEventStreamStop(stream);
+                fse::FSEventStreamSetDispatchQueue(stream, ptr::null_mut());
+                fse::FSEventStreamInvalidate(stream);
+                dispatch2::ffi::dispatch_release(self.queue as dispatch_object_t);
+                fse::FSEventStreamRelease(stream);
+            }
+            None => unsafe {
+                dispatch2::ffi::dispatch_release(self.queue as dispatch_object_t);
+            }
+        };
+    }
+}
+
+struct ContextInfo {
+    handler: Box<dyn Fn(Kind, &String)>,
+}
+
+const FLAGS : fse::FSEventStreamCreateFlags
+        = fse::kFSEventStreamCreateFlagNoDefer
+        | fse::kFSEventStreamCreateFlagWatchRoot
+        | fse::kFSEventStreamCreateFlagFileEvents;
+
+extern "C" fn release_context(info: *mut c_void) {
+  let ctx_ptr = info as *mut ContextInfo;
+  unsafe{ drop(Box::from_raw( ctx_ptr)); }
+}
+
+extern "C" fn callback(
+    _stream_ref: fse::FSEventStreamRef,
+    info: *mut c_void,
+    num_events: usize,
+    event_paths: *mut c_void,
+    event_flags: *const fse::FSEventStreamEventFlags,
+    _event_ids: *const fse::FSEventStreamEventId,
+) {
+    let info = unsafe{ &mut *(info as *mut ContextInfo) };
+    let handler = info.handler.as_ref();
+
+    let event_paths = event_paths as *const *const c_char;
+    for i in 0..num_events {
+        let path = unsafe { CStr::from_ptr(*event_paths.add(i)).to_str().unwrap().to_string() };
```

Comment:
add a TODO here to see if we can avoid copying the `CStr` data into a rust string, as we know we'll be copying it to a jni string later on..